src/ui/edit_field.c

```diff
@@ -20,14 +20,270 @@ struct Edit_field
+static void kill_word(Edit_field *edit_field);
+static void backward_kill_word(Edit_field *edit_field);
+
+static void handle_keydown(Edit_field *edit_field, const SDL_Event *event);
+static void handle_keydown_alt(Edit_field *edit_field, const SDL_Event *event);
+static void handle_keydown_ctrl(Edit_field *edit_field, const SDL_Event *event);
+
+static void edit_field_insert_char(Edit_field *edit_field, char c)
+{
+    if (edit_field->buffer_size >= BUFFER_CAPACITY) {
+        return;
+    }
+
+    for (int64_t i = (int64_t) edit_field->buffer_size - 1; i >= (int64_t) edit_field->cursor; --i) {
+        edit_field->buffer[i + 1] = edit_field->buffer[i];
+    }
+
+    edit_field->buffer[edit_field->cursor++] = c;
+    edit_field->buffer[++edit_field->buffer_size] = 0;
+}
+
+// See: https://www.gnu.org/software/emacs/manual/html_node/emacs/Moving-Point.html
+// For an explanation of the naming terminology for these helper methods
+
+static bool is_emacs_word(char c)
+{
+#define IN_RANGE(start, end) (c >= start && c <= end)
+
+    // Word syntax table retrieved from Fundamental Mode, "C-h s"
+    // (This is not the complete syntax table)
+    return IN_RANGE('$', '%')
+        || IN_RANGE('0', '9')
+        || IN_RANGE('A', 'Z')
+        || IN_RANGE('a', 'z')
+        ;
+
+#undef IN_RANGE
+}
+
+static void forward_char(Edit_field *edit_field)
+{
+    // "C-f" or "<RIGHT>"
+    if (edit_field->cursor < edit_field->buffer_size) {
+        edit_field->cursor++;
+    }
+}
+
+static void backward_char(Edit_field *edit_field)
+{
+    // "C-b" or "<LEFT>"
+    if (edit_field->cursor > 0) {
+        edit_field->cursor--;
+    }
+}
+
+static void move_beginning_of_line(Edit_field *edit_field)
+{
+    // "C-a" or "<Home>"
+    edit_field->cursor = 0;
+}
+
+static void move_end_of_line(Edit_field *edit_field)
+{
+    // "C-e" or "<End>"
+    edit_field->cursor = edit_field->buffer_size;
+}
+
+static void forward_word(Edit_field *edit_field)
+{
+    // "M-f" or "C-<RIGHT>" or "M-<RIGHT>"
+    while (true) {
+        forward_char(edit_field);
+        if (edit_field->cursor >= edit_field->buffer_size) {
+            break;
+        }
+
+        char current = edit_field->buffer[edit_field->cursor];
+        char preceeding = edit_field->buffer[edit_field->cursor - 1];
+        if (!is_emacs_word(current) && is_emacs_word(preceeding)) {
+            // Reached the end of the current word
+            break;
+        }
+    }
+}
+
+static void backward_word(Edit_field *edit_field)
+{
+    // "M-b" or "C-<LEFT>" or "M-<LEFT>"
+    while (true) {
+        backward_char(edit_field);
+        if (edit_field->cursor == 0) {
+            break;
+        }
+
+        char current = edit_field->buffer[edit_field->cursor];
+        char preceeding = edit_field->buffer[edit_field->cursor - 1];
+        if (is_emacs_word(current) && !is_emacs_word(preceeding)) {
+            // Reached the start of the current word
+            break;
+        }
+    }
```

Comment:
Why can't this be something like `while(cursor >= 0 && is_emacs_word(buffer[cursor])) { --cursor; }`?  Why does it have to be so over-complicated?